### DISCUSSION TITLE
assigning a type-specific function to a generic name at runtime

### DISCUSSION THREAD

**Discussion by dhduvall** [microsoft/pyright#6935] [Q&A]
I've got code with this structure. It's assigning a series of handle-agnostic functions from a matching series of handle-specific functions based on a runtime decision (here from an environment variable). I'd prefer to do this over a method where I define one function `func` that can take multiple types and make the decision every time it's called. Is it possible, short of typing `handle` as `Any`, to get pyright to understand what I'm doing here (it seems like that might be difficult)? Or is there a better way to structure this code?

```python
  import os

  class HandleA():
      pass

  class HandleB():
      pass

  def func_a(handle: HandleA) -> None:
      print("called func_a")

  def func_b(handle: HandleB) -> None:
      print("called func_b")

  t = os.environ.get("TYPE", "a")

  if t == "a":
      func = func_a
  elif t == "b":
      func = func_b
  else:
      raise NotImplementedError(f"Don't know what type {t} is")

  if t == "a":
      handle = HandleA()
  elif t == "b":
      handle = HandleB()
  else:
      raise NotImplementedError(f"Don't know what type {t} is")

  # Argument of type "HandleA | HandleB" cannot be assigned to parameter "handle" of type "HandleA" in function "func_a"
  # Argument of type "HandleA | HandleB" cannot be assigned to parameter "handle" of type "HandleB" in function "func_b"
  func(handle)
```

**Comment by erictraut** [comment]
This would be a more typical way to handle this situation:

```python
class HandleA:
    pass

class HandleB:
    pass

Handle = HandleA | HandleB

def process_handle(handle: Handle) -> None:
    match handle:
        case HandleA():
            print("called func_a")
        case HandleB():
            print("called func_b")

def func(handle: Handle):
    process_handle(handle)
```

**Reply by dhduvall** [reply]
So make the decision in each call, but by checking the type of `Handle`?

Does this create handle objects as it processes each `case` statement?

**Reply by erictraut** [reply]
> Does this create handle objects as it processes each `case` statement?

No, these are patterns that the `match` statement is comparing against. This code is equivalent to:

```python
def process_handle(handle: Handle) -> None:
    if isinstance(handle, HandleA):
        print("called func_a")
    elif isinstance(handle, HandleB):
        print("called func_b")
```

**Reply by dhduvall** [reply]
I haven't programmed in Python for a few years, so the details of `match` are new to me.

Not that I really need to be worried about performance in this case, but I still don't like the idea of doing this check on every call rather than once at the beginning of execution. Is there really no way to maintain that property?

Thanks for the help.

**Reply by erictraut** [reply]
The way you've written the code, you have two separate variables (`handle` and `func`) that can each have one of two different types. A type checker doesn't know that the two variables are "entangled" such that if the type of `handle` is `HandleA` then it can assume that `func` is always `func_a`, etc. 

If you allocate `handle` only once (as you're doing in the example code), you can bind that instance of `handle` with the associated function using `functools.partial`. Then you can simply call `func` with no arguments any time you need to execute it.

```python
if t == "a":
    func = partial(func_a, HandleA())
elif t == "b":
    func = partial(func_b, HandleB())
else:
    raise NotImplementedError(f"Don't know what type {t} is")

func()
```
